tools/monster.c

```diff
@@ -1349,141 +1276,42 @@ char* smt_ternary(char* opt, char* op1, char* op2, char* op3) {
+uint64_t  get_depth(uint64_t* node)        { return             *(node + 4); }
 
-  entry = smalloc(2 * SIZEOFUINT64STAR);
-
-  *(entry + 0) = (uint64_t) mergeable_contexts;
-  *(entry + 1) = (uint64_t) context;
-
-  mergeable_contexts = entry;
-}
-
-uint64_t* get_mergeable_context() {
-  uint64_t* head;
-
-  if (mergeable_contexts == (uint64_t*) 0)
-    return (uint64_t*) 0;
-
-  head = mergeable_contexts;
-  mergeable_contexts = (uint64_t*) *(head + 0);
-
-  return (uint64_t*) *(head + 1);
-}
-
-void add_waiting_context(uint64_t* context) {
-  uint64_t* entry;
-
-  entry = smalloc(2 * SIZEOFUINT64STAR);
-
-  *(entry + 0) = (uint64_t) waiting_contexts;
-  *(entry + 1) = (uint64_t) context;
-
-  waiting_contexts = entry;
-}
-
-uint64_t* get_waiting_context() {
-  uint64_t* head;
-
-  if (waiting_contexts == (uint64_t*) 0)
-    return (uint64_t*) 0;
-
-  head = waiting_contexts;
-  waiting_contexts = (uint64_t*) *(head + 0);
-
-  return (uint64_t*) *(head + 1);
-}
+void set_parent_node(uint64_t* node, uint64_t* parent)  { *(node + 0) = (uint64_t) parent; }
+void set_children(uint64_t* node, uint64_t* children)   { *(node + 1) = (uint64_t) children; }
+void set_sibling(uint64_t* node, uint64_t* sibling)     { *(node + 2) = (uint64_t) sibling; }
+void set_node_address(uint64_t* node, uint64_t address) { *(node + 3) = address; }
+void set_depth(uint64_t* node, uint64_t depth)          { *(node + 4) = depth; }
 
 void merge(uint64_t* active_context, uint64_t* mergeable_context, uint64_t location) {
-  uint64_t callstack_comparison;
 
   // do not merge if merging is disabled
-  if (merge_enabled == 0) {
-    if (current_mergeable_context != (uint64_t*) 0) {
-      add_mergeable_context(current_mergeable_context);
-      current_mergeable_context = (uint64_t*) 0;
-    }
-
-    return;
-  }
-
-  if (active_context == mergeable_context) {
-    current_mergeable_context = get_mergeable_context();
-
-    if (current_mergeable_context != (uint64_t*) 0)
-      if (pc == get_pc(current_mergeable_context))
-        merge(active_context, current_mergeable_context, pc);
-    return;
-  }
-
-  callstack_comparison = compare_call_stacks(active_context, mergeable_context);
-
-  if (callstack_comparison == 2) { // mergeable context has longer call stack
-    throw_exception(EXCEPTION_SYMBOLICRECURSION, 0);
-    return;
-  } else if (callstack_comparison != 0) { // call stacks are not equal
-    if (current_mergeable_context != (uint64_t*) 0) {
-      add_mergeable_context(current_mergeable_context);
-      current_mergeable_context = (uint64_t*) 0;
-    }
-
+  if (merge_enabled == 0)
     return;
-  }
 
   print("; merging two contexts at ");
   print_code_context_for_instruction(location);
 
   if (debug_merge)
-    printf2(" -> active context: %u, mergeable context: %u", (char*) active_context, (char*) mergeable_context);
+    printf2(" -> active context: %d, mergeable context: %d", (char*) active_context, (char*) mergeable_context);
```

Comment:
This should remain %u, no?

Reply:
I must have overlooked that change. I actually think both would work fine, because it's just about identifying contexts for debugging. I changed it to %u.